added send_contact, send_chat_action and added reply_markup for send_…

TheUberCatman opened the pull request:
here are my created methods

voider1 commented:
Could you please tell me first what you're exactly adding? And work on another branch than master in the future please.

TheUberCatman commented:
I added some bot messaging methods,

send_contact: sending contacts by creating a Contact object and passing that to the send_contact method.

send_chat_action: send chat_action to the chat that tells that the bot is typing or sending something. chat_actions can be found in objects.

added reply_markup to send_message method in messaging. it basically is button that can act as a link or callback.

voider1 commented:
Alright, nice work! There are some things which need to be changed before I can merge the PR. It's not yet up to the standards of the rest of the library.

voider1 commented on src/bot/messaging.rs:
```diff
@@ -61,6 +71,71 @@ impl Bot {
+        if message_json.as_bool().unwrap() {
+            Ok(true)
+        } else {
+            Ok(false)
```
Why is it returning `false`? The docs say it returns `true` on success, you should try to return the contents from the JSON response.

voider1 commented on src/bot/messaging.rs:
```diff
@@ -61,6 +71,71 @@ impl Bot {
+        println!("{:?}", &data);
+
+        let rjson: Value = check_for_error(data.json()?)?;
+        let message_json = rjson.get("result").ok_or(JsonNotFound)?;
```
I don't think `message_json` is a semantically correct name for this variable binding. This action will not return `Message` object in JSON form.

voider1 commented on src/bot/messaging.rs:
```diff
@@ -61,6 +71,71 @@ impl Bot {
+        let reply_to_message_id: &str = &reply_to_message_id.map(|i| i.to_string())
+            .unwrap_or("None".to_string());
+        let reply_markup =
+            if reply_markup.is_some() {
```
If you're going to unwrap it anyways, change it to an if-let statement.

voider1 commented on src/bot/messaging.rs:
```diff
@@ -61,6 +71,71 @@ impl Bot {
+            ("reply_markup", &reply_markup)];
+        //println!("{:?}", params);
+        let mut data = self.client.post(&url).form(&params).send()?;
+        println!("{:?}", &data);
```
Remove debug `println!`.

voider1 commented on src/bot/messaging.rs:
```diff
@@ -61,6 +71,71 @@ impl Bot {
+        println!("{:?}", &data);
+
+        let rjson: Value = check_for_error(data.json()?)?;
+        println!("{:?}", &rjson);
```
Remove debug `println!`.

voider1 commented on src/bot/messaging.rs:
```diff
@@ -61,6 +71,71 @@ impl Bot {
+        println!("{:?}", &rjson);
+        let message_json = rjson.get("result").ok_or(JsonNotFound)?;
+        //println!("{:?}", message_json);
+        if message_json.as_bool().unwrap_or(false) {
```
Please look at my other methods and refer to [this part](https://core.telegram.org/bots/api#sendcontact) of the Telegram bot API docs. It should return the sent message on success, otherwise it should throw an error. Returning `Ok` means something has gone correct. This isn't the case. Like the Rust docs say: "Result is a type that represents either success (Ok) or failure (Err)."

voider1 commented on src/objects/inline_keyboard.rs:
```diff
@@ -0,0 +1,39 @@
+	switch_inline_query_current_chat: Option<String>,
+}
+
+impl Markup {
```
The name of this in the Telegram bot API is "InlineKeyboardMarkup", please try to keep the names as close as possible.

voider1 commented on src/objects/inline_keyboard.rs:
```diff
@@ -0,0 +1,39 @@
+}
+
+#[derive(Serialize, Debug)]
+pub struct Button {
```
The name of this in the Telegram bot API is "InlineKeyboardButton", please try to keep the names as close as possible.

TheUberCatman commented on src/bot/messaging.rs:
```diff
@@ -61,6 +71,71 @@ impl Bot {
     }
 
+    /// API call whick will send the requested chat action to your bot.
+    pub fn send_chat_action(&self, chat_id: &i64, action: &ChatAction) -> Result<bool> {
```
let me know what you think is best

voider1 replied:
Put it in mod.rs.

voider1 commented on src/bot/media.rs:
```diff
@@ -0,0 +1,44 @@
+use objects::{InlineKeyboardMarkup, Contact};
+use bot::Bot;
+
+impl Bot {
```
You can remove this impl and move all the methods into mod.rs.

voider1 commented on src/bot/messaging.rs:
```diff
@@ -61,6 +71,21 @@ impl Bot {
+        let path = ["sendChatAction"];
+        let url = ::construct_api_url(&self.bot_url, &path);
+        let params = [("chat_id", chat_id.to_string()),
+            ("action", action)];
```
Ident this please.

voider1 commented on src/objects/inline_keyboard.rs:
```diff
@@ -0,0 +1,39 @@
+}
+
+#[derive(Serialize, Debug)]
+pub struct InlineKeyboardButton {
```
Please fix indentation in this file.

voider1 commented on tests/test_chat_action.rs:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    #[test]
+    fn test_url_chat_actions() {
```
You can move this test to test_bot.rs.

voider1 commented on tests/test_inline_keyboard.rs:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    #[test]
+    fn test_url_inline_keyboard() {
```
You can move this test to bot.rs too.

voider1 commented on tests/test_send_contact.rs:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    #[test]
+    fn test_send_contact() {
```
This test can be moved to test_bot.rs too.

voider1 commented on src/bot/messaging.rs:
```diff
@@ -24,13 +26,21 @@ impl Bot {
         let reply_to_message_id: &str = &reply_to_message_id.map(|i| i.to_string())
             .unwrap_or("None".to_string());
+
+        let reply_markup: &str = &if let Some(r) = reply_markup {
```
Or maybe use:

```rust
let reply_markup = &reply_markup.map(|r| serde_json::to_string(r).unwrap_or("".to_string()));
```

voider1 commented on src/bot/mod.rs:
```diff
@@ -89,4 +92,112 @@ impl Bot {
+        let url = ::construct_api_url(&self.bot_url, &path);
+        let mut data = self.client.post(&url).form(&params).send()?;
+        let rjson: Value = check_for_error(data.json()?)?;
+        let result = rjson.get("result").unwrap().as_bool().unwrap();
```
Could you please add proper error handling here? The response you get back should be treated as unreliable, you know what you expect, but what you expect isn't always what you get.

voider1 commented on src/bot/mod.rs:
```diff
@@ -89,4 +92,112 @@ impl Bot {
+                        chat_id: &i64,
+                        contact: &Contact,
+                        disable_notification: Option<&bool>,
+                        reply_to_message_id: Option<&i32>,
```
I'd prefer more consistency, the `send_message` method has `reply_to_message_id` defined as an `Option<&i64>`. All API interactions which use integers except from `get_updates` use the `&i64`.

voider1 commented on src/bot/mod.rs:
```diff
@@ -89,4 +92,112 @@ impl Bot {
+        let disable_notification: &str = &disable_notification.unwrap_or(&false).to_string();
+        let reply_to_message_id: &str = &reply_to_message_id.map(|i| i.to_string())
+            .unwrap_or("None".to_string());
+        let reply_markup = &reply_markup.map(|r| serde_json::to_string(r).unwrap_or("".to_string())).unwrap_or("".to_string());
```
This is very unreadable, could you try to cut it down? Maybe try something like:

```rust
reply_markup.and_then(|r| serde_json::to_string(r).ok())).unwrap_or("".to_string());
```

I haven't tested it.

voider1 commented on src/bot/mod.rs:
```diff
@@ -89,4 +92,112 @@ impl Bot {
+
+        let path = ["sendContact"];
+        let params = [("chat_id", chat_id),
+                      ("phone_number", &contact.phone_number),
```
I'd prefer if you didn't do this and keep the same style as I have, define your values above the `params`, don't do referencing and unwrapping inside the tuple, it's asking for trouble.

voider1 commented on src/bot/mod.rs:
```diff
@@ -89,4 +92,112 @@ impl Bot {
+        let mut data = self.client.post(&url).form(&params).send()?;
+        let rjson: Value = check_for_error(data.json()?)?;
+        let message_json = rjson.get("result").ok_or(JsonNotFound)?;
+        println!("{:?}", message_json);
```
Could you please remove all debug `println!` statements?

voider1 commented on src/bot/parse_mode.rs:
```diff
@@ -14,3 +14,4 @@ pub fn get_parse_mode(parse_mode: &ParseMode) -> String {
     }
     .to_string()
 }
+
```
Could you please remove this whiteline?

voider1 commented on src/objects/call_back_query.rs:
```diff
@@ -0,0 +1,13 @@
+
+#[derive(Clone, Deserialize, Debug)]
+pub struct CallBackQuery {
+    id: String,
```
Are you sure the user doesn't need to access any of these values?

voider1 commented:
Also a more general comment, run RustFmt on your files. =)

voider1 commented:
Good job, looks good! Can you resolve the conflicts? I'll merge it after that.

voider1 commented:
Could you please run [RustFmt](https://github.com/rust-lang-nursery/rustfmt) on your files?

voider1 commented on src/objects/call_back_query.rs:
```diff
@@ -0,0 +1,13 @@
+use objects::user::User;
+
+#[derive(Clone, Deserialize, Debug)]
+pub struct CallBackQuery {
```
Please write documentation.

voider1 commented on src/objects/contact.rs:
```diff
@@ -0,0 +1,7 @@
+#[derive(Clone, Serialize, Deserialize, Debug)]
+pub struct Contact {
```
Please write documentation.

voider1 commented on src/objects/inline_keyboard.rs:
```diff
@@ -0,0 +1,39 @@
+use std::option::Option;
+
+#[derive(Serialize, Debug)]
+pub struct InlineKeyboardMarkup {
```
Please write documentation.

voider1 commented on src/objects/inline_keyboard.rs:
```diff
@@ -0,0 +1,39 @@
+}
+
+#[derive(Serialize, Debug)]
+pub struct InlineKeyboardButton {
```
Please write documentation.

voider1 commented on src/bot/chat_action.rs:
```diff
@@ -0,0 +1,24 @@
+/// The ChatActions to send.
+pub enum ChatAction {
```
Please write documentation.

voider1 commented on src/objects/call_back_query.rs:
```diff
@@ -0,0 +1,14 @@
+use objects::message::Message;
+use objects::user::User;
+
+/// represents an incoming callback query from a callback button in an inline keyboard.
```
Please use capital letters at the beginning of a sentence. When referring to another type, like "inline keyboard", use the type like `InlineKeyboardMarkup`.

voider1 commented on src/objects/contact.rs:
```diff
@@ -0,0 +1,8 @@
+/// represents a phone contact.
```
Please use a capital letter at the beginning of a sentence.

voider1 commented on src/objects/inline_keyboard.rs:
```diff
@@ -0,0 +1,43 @@
+use std::option::Option;
+
+/// represents an inline keyboard that appears right next to the message it belongs to.
```
Please us a capital letter at the beginning of the sentence. When referring to other types do it like `Message`.

voider1 commented on src/objects/inline_keyboard.rs:
```diff
@@ -0,0 +1,43 @@
+    pub inline_keyboard: Vec<Vec<InlineKeyboardButton>>,
+}
+
+/// represents one button of an inline keyboard. You must use exactly one of the optional fields.
```
Same as the others.

voider1 commented on src/bot/chat_action.rs:
```diff
@@ -0,0 +1,24 @@
+/// The ChatActions to send.
```
Please change this to

```rust
/// Represents each chat action.
```

voider1 commented on tests/test_bot.rs:
```diff
@@ -73,7 +81,53 @@ mod tests {
+        let bot_url = [BASE_URL, &token].concat();
+        let bot = Bot::new(bot_url).unwrap();
+
+        let mut total = Vec::<Vec<InlineKeyboardButton>>::new();
```
I think that `buttons` would be a better name than `total` for this variable binding.

voider1 commented on tests/test_bot.rs:
```diff
@@ -73,7 +81,53 @@ mod tests {
+        let bot_url = [BASE_URL, &token].concat();
+        let bot = Bot::new(bot_url).unwrap();
+
+        let message = bot.send_chat_action(&chat_id, &ChatAction::FindLocation);
```
I think `success` would be a better name than `message`, because `send_chat_action` returns a boolean which indicates success instead of a message.

voider1 commented:
Alright, good work on this PR. I will merge and close it. Thank you for your contribution. =)

voider1 commented on src/objects/inline_keyboard.rs:
```diff
@@ -0,0 +1,43 @@
+    pub inline_keyboard: Vec<Vec<InlineKeyboardButton>>,
+}
+
+/// Represents one button of an `InlineKeyboardMarkup` You must use exactly one of the optional fields.
```
You -> you

voider1 replied:
Or add a dot before "You".

voider1 commented:
Good job on this PR! I will merge and close it. Thanks for your contribution.
